### DISCUSSION TITLE
Get small amount of native data from CQL?

### DISCUSSION THREAD

**Discussion by amyworrall** [ricomariani/CG-SQL-author#78] [Q&A] [answered]
I'm writing an undo manager. Associated with each sqlite database I've got a native object that represents the entire database connection. On this native object I've got a boolean property, `is_redoing`, which shows whether a redo operation is currently in progress.

I need to access this boolean from within my CQL sprocs, to make sure I'm not logging new undo information when a redo operation is going on.

If I define a native C function, I can call this from within CQL. But in that native code, I wouldn't have any idea _which_ database was making the request. My app may need to hold multiple database connections (to different sqlite databases) at once, and each one should have their own undo system.

Whatever method I use here, I'll need to be able to call it from within a sqlite trigger.

Any suggestions on the best way to go? A vtable would work but seems very heavyweight for reading one bool… is there anything easier?

**Comment by amyworrall** [comment]
Addendum: I'd like to be able to _pull_ data from the native side. If I were to _push_ it (i.e. trigger something from the native side whenever is_redoing changes), it'd be easier, I could just shove it in a temp table. But that seems less elegant (doing more work more often), and there may be other situations in the future where I absolutely have to pull the data.

**Comment by ricomariani** [comment, marked as answer]
From cql, instead of declaring a function, declare a procedure.  Implement the proc in C. Use an out bool not null arg.  You will be passed the database pointer.

There are many examples like this in cqlrt.

Return sqlite_ok for success.

Functions are for things that don't use the database and can't fail.

Any procedure with a single trailing out arg can be called like a function using the out as the return value. This is just sugar but it's cleaner looking.

Check out the fib example in the docs.

When you declare the procedure make sure you add 'using transaction' so that it will get the db pointer.  Else you get a void signature.

Fwiw There's a chapter on this stuff in the docs.



Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
